p2p/dhtdiscovery/discovery.go

```diff
@@ -444,100 +487,116 @@ func (d *Discovery) advertiseNS(ctx context.Context, ns string, active func() bo
-
-	return nil
+func (d *Discovery) discoverPeers(ctx context.Context) {
+	peerCh := d.findPeersContinuously(ctx, discoveryNS, d.h.NeedPeerDiscovery)
+	d.connBackoff.Connect(ctx, peerCh)
 }
 
 func (d *Discovery) discoverRelays(ctx context.Context) {
-	for p := range d.findPeersContinuously(ctx, relayNS) {
+	for p := range d.findPeersContinuously(ctx, relayNS, nil) {
 		if len(p.Addrs) != 0 {
 			d.logger.Debug("found relay candidate", zap.Any("p", p))
 			select {
-			case d.relayCh <- p:
 			case <-ctx.Done():
 				return
+			case d.relayCh <- p:
 			}
 		}
 	}
 }
 
-func (d *Discovery) findPeersContinuously(ctx context.Context, ns string) <-chan peer.AddrInfo {
-	r := make(chan peer.AddrInfo)
+func (d *Discovery) findPeersContinuously(
+	ctx context.Context,
+	ns string,
+	active func() bool,
+) <-chan peer.AddrInfo {
+	peerCh := make(chan peer.AddrInfo)
 	d.eg.Go(func() error {
-		defer close(r)
-		var peerCh <-chan peer.AddrInfo
+		defer close(peerCh)
 		for {
-			if peerCh == nil {
-				var err error
-				peerCh, err = d.disc.FindPeers(ctx, ns)
-				if err != nil {
-					d.logger.Error("error finding relay peers", zap.Error(err))
-					select {
-					case <-ctx.Done():
-						return nil
-					case <-time.After(d.findPeersRetryDelay):
-					}
-					peerCh = nil
-					continue
-				}
+			cont, err := d.findPeers(ctx, ns, active, discRecheckInterval, peerCh)
+			if err != nil {
+				d.logger.Error("error finding relay peers", zap.String("ns", ns), zap.Error(err))
 			}
-
+			if !cont {
+				return nil
+			}
+			d.logger.Debug("pausing discovery", zap.String("ns", ns),
+				zap.Duration("duration", d.findPeersRetryDelay))
 			select {
 			case <-ctx.Done():
 				return nil
-			case p, ok := <-peerCh:
-				if !ok {
-					peerCh = nil
-					select {
-					case <-ctx.Done():
-						return nil
-					case <-time.After(d.findPeersRetryDelay):
-					}
-					continue
-				}
-				select {
-				case r <- p:
-				case <-ctx.Done():
-				}
+			case <-time.After(d.findPeersRetryDelay):
 			}
 		}
 	})
-	return r
+	return peerCh
+}
+
+func (d *Discovery) findPeers(
+	ctx context.Context,
+	ns string,
+	active func() bool,
+	checkInterval time.Duration,
+	out chan<- peer.AddrInfo,
+) (cont bool, err error) {
+	if ctx.Err() != nil {
+		return false, err
+	}
+	if active != nil && !active() {
+		d.logger.Debug("discovery not active", zap.String("ns", ns))
+		return true, nil
+	}
+	d.logger.Debug("started finding peers", zap.String("ns", ns))
+	defer d.logger.Debug("done finding peers", zap.String("ns", ns))
+	childCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	peerCh, err := d.disc.FindPeers(childCtx, ns)
+	if err != nil {
+		return !errors.Is(err, context.Canceled), err
+	}
+	var tickerCh <-chan time.Time
+	if active != nil {
+		ticker := time.NewTicker(checkInterval)
+		tickerCh = ticker.C
+		defer ticker.Stop()
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return false, nil
+		case <-tickerCh:
+			if !active() {
+				return true, nil
+			}
```

Comment:
It took me a while to understand that `active` is never `nil` here. Please, consider adding a comment explaining it.
```suggestion
		case <-tickerCh:
			// Note: ticker is not started if active is nil.
			if !active() {
				return true, nil
			}
```

Reply:
Makes sense, thanks